flow/hooks/git_workspace_tracking.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2018 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from __future__ import print_function
+import os
+import json
+import logging
+from collections import defaultdict
+
+import git
+
+from .git_util import collect_metadata_with_git as collect_metadata
+
+
+logger = logging.getLogger('git_tracking')
+
+GIT_COMMIT_MSG = """{title}
+
+*This commit was auto-generated.*
+"""
+
+
+_WARNINGS = defaultdict(set)
+
+
+def _ignore_hidden_files(root):
+    "Do not commit hidden files to the git repository."
+    with open(os.path.join(root, '.gitignore'), mode='a+') as file:
+        file.seek(0)
+        if '.*\n' not in file.readlines():
+            file.write('.*\n')
+
+
+def _get_or_init_git_repo(root):
+    "Return the git repository for root and initialize it if necessary."
+    try:
+        return git.Repo(root)
+    except git.exc.InvalidGitRepositoryError:
+        logger.info("Initializing git repo for '{}'".format(root))
+        _ignore_hidden_files(root)
+        return git.Repo.init(root)
+
+
+def _commit(repo, title):
+    try:
+        repo.git.commit('-m', GIT_COMMIT_MSG.format(title=title))
+    except git.exc.GitCommandError as error:
+        if "nothing to commit, working tree clean" in str(error):
+            pass
+        else:
+            raise
```

Comment:
This will also raise an error if there are *untracked* files in the repository (see [here](https://bitbucket.org/glotzer/signac-flow/pull-requests/35/execution-hooks/diff#comment-72666460))

Reply:
Shouldn't be an issue, because we should automatically add all files (see response below).